testonly/hammer/record.go

```diff
@@ -28,48 +30,99 @@ const (
 	smrCount = 30
 )
 
-// smrStash provides thread-safe access to an ordered, bounded queue of previously
-// witnessed SMRs
-type smrStash struct {
-	mu sync.RWMutex
+// gossipHub allows details of what has been written to and read from the Map
+// to be shared between different workers running in the same process. The same
+// interface could be implemented by something which shared these details with
+// a remote process via gRPC if that was desirable.
+// The goal is thatVersionedMapContents can be subsumed into this class - there
+// should be no reason to address it directly. For now, this class coordinates
+// writes, reads, and local state such that VersionedMapContents always contains
+// revisions which are no further ahead than the most recent SMR published by the
+// Map.
+type gossipHub struct {
+	contents *testonly.VersionedMapContents
+
+	mu            sync.RWMutex
+	pendingWrites map[uint64][]*trillian.MapLeaf
 
 	// SMRs are arranged from later to earlier (so [0] is the most recent), and the
 	// discovery of new SMRs will push older ones off the end. No SMR for a revision
 	// will be stored more than once.
-	smr [smrCount]*types.MapRootV1
+	smrs [smrCount]*types.MapRootV1
+}
+
+func newGossipHub() *gossipHub {
+	return &gossipHub{
+		contents:      &testonly.VersionedMapContents{},
+		pendingWrites: make(map[uint64][]*trillian.MapLeaf),
+	}
+}
+
+func (g *gossipHub) getLastReadRev() (rev uint64, found bool) {
+	if lastRoot := g.previousSMR(0); lastRoot != nil {
+		return lastRoot.Revision, true
+	}
+	return 0, false
+}
+
+// proposeLeaves should be called *before* writing a new map revision. The revision
+// will be available to readers only once advertiseSMR has been called with a root
+// that is >= the revision of this write.
+func (g *gossipHub) proposeLeaves(rev uint64, leaves []*trillian.MapLeaf) error {
+	if head, found := g.getLastReadRev(); found && rev <= head {
```

Comment:
I threw this away. It was only trying to be defensive and catch a programming error, but it complicates the readability of the code. If this check would have failed here, it will fail when the writer proceeds to write this proposal to the map.